plugins/actuators/vllm_performance/ado_actuators/vllm_performance/env_manager.py

```diff
@@ -125,68 +139,115 @@ def get_environment(
         :param increment_usage: increment usage flag
         :return: environment state
         """
+        print(
+            f"getting environment for model {model}, currently {self.active_environments} deployments"
+        )
 
-        env = self.environments.get(definition, None)
+        # check if there's an existing free environment satisfying the request
+        env = self.get_matching_free_environment(definition)
         if env is None:
-            if len(self.environments) >= self.max_concurrent:
+            if self.active_environments >= self.max_concurrent:
                 # can't create more environments now, need clean up
-                available = False
-                for key, env in self.environments.items():
-                    if env.in_use == 0:
-                        available = True
-                        start = time.time()
-                        try:
-                            self.manager.delete_service(k8s_name=env.k8s_name)
-                            self.manager.delete_deployment(k8s_name=env.k8s_name)
-                        except ApiException as e:
-                            logger.error(f"Error deleting deployment or service {e}")
-                        del self.environments[key]
-                        logging.info(
-                            f"deleted environment {env.k8s_name} in {time.time() - start} sec. "
-                            f"Environments length {len(self.environments)}"
+                if len(self.free_environments) > 0:
+                    # There are unused environments, let's evict one
+
+                    # Gets the oldest env in the dict
+                    venv_to_evict = self.free_environments[0]
+                    try:
+                        self.manager.delete_service(k8s_name=venv_to_evict.k8s_name)
+                        self.manager.delete_deployment(k8s_name=venv_to_evict.k8s_name)
+                        logger.info(
+                            f"deleted environment {venv_to_evict.k8s_name}. "
+                            f"Active environments {self.active_environments}"
                         )
-                        time.sleep(3)
-                        break
-                if not available:
+                    except ApiException as e:
+                        logger.error(f"Error deleting deployment or service {e}")
```

Comment:
Here you delete everything in one single try and do not raise anything in case of errors - is it fine?

Reply:
it is actually not fine, good catch.